### PR TITLE
dockerised or containerised fabric

### PR DESCRIPTION
**Pull request by fahadshery**
## What this Pull Request (PR) does
I added the `docker` folder with the required `DockerFile` etc. to containerise `fabric application`.

## Related issues
`closes #327`.

### CODE REVIEW

**Comment by danielmiessler** [comment]
@xssdoctor what do you think on this one?

File: docker/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+Replace < latest tag > with any version identifier, like v1.0 from fabric project.
+
+The . at the end of the command indicates that you're building the image using `DockerFile` available in the current directory.
+
+### Running the Container:
+
+After building the image, run the container using the following command:
+
+```
+$ docker run --rm -i -t -d --name fabric -v ./fabric/:/home/appuser/.config/fabric --network traefik fabic:latest
```
Comment (LaBalaDeOro) [inline docker/README.md]: Typo: `fabic:latest` should be `fabric:latest`

**Comment by bijeebuss** [comment]
+1 on this. This allows me to run this without installing a bunch of stuff on my computer

**Comment by bijeebuss** [comment]
actually apparently I can just do this and it saves the install of fabic 
```
alias gocontainer='docker run -it --pull always --rm -v $HOME:/root -v $(pwd):/go -w /go golang:latest'
```

**Comment by nloadholtes** [comment]
Quick note: following the instructions in the `README.md` did not work on an ubuntu system: the dockerfile is named `DockerFile` but my docker was looking for `Dockerfile`. (Running with `-f DockerFile` let it build, but most dockerfiles I've seen only have the `D` capitalized.)

Thank you for making this PR!

File: docker/DockerFile
Comment (nloadholtes) [inline docker/DockerFile:1]: Quick note: following the instructions in the README.md did not work on an ubuntu system: the dockerfile is named `DockerFile` but my docker was looking for `Dockerfile`. (Running with -f DockerFile let it build, but most dockerfiles I've seen only have the `D` capitalized.)

Thank you for making this PR!

**Review by brent-dev — APPROVED** [review]

**Comment by eugeis** [comment]
Thank you for the PR.

Please also check those 3 points.

1) Docker image for production
You image will work for sure. The best practice for production is to have small docker images only with app it self, without not needed dependencies. To run fabric we don't need the Go dev. environment, but only the binary. In order to have such a image, usually you use a multi stage Docker file. Please check the [example](https://docs.docker.com/build/building/multi-stage/?_gl=1*ygl7ff*_gcl_au*MjczODYyODg3LjE3MjE2NTk1ODY.*_ga*NDE5NzI3MjI1LjE3MTc0NTA2MzE.*_ga_XJWPQMJYHQ*MTcyNDk1NDA0NC4yMy4xLjE3MjQ5NTQ0MDAuNTguMC4w) here and adapt the Dockerfile.

2. Please clean up docker-compose file. Remove all commented out lines and the traefic network.

3. Please add Github Action to build and publish Docker image to Github Registry, see e.g. [here](https://docs.github.com/en/actions/sharing-automations/creating-actions/creating-a-docker-container-action)

**Comment by danielmiessler** [comment]
Love it.

**Comment by brauliodiasribeiro** [comment]
Someone help me install Fabric AI in a docker on Unraid?
I don't know how to build this docker in Unraid because it asks for the image path.

**Comment by yousecjoe** [comment]
> Someone help me install Fabric AI in a docker on Unraid? I don't know how to build this docker in Unraid because it asks for the image path.

It would be best to learn to build the image yourself. Unless I am mistaken, based on reading this PR, you still need to build the image for yourself because the choice to automate the build and upload to GHCR was pushed back on the core dev, and then there was some confusing statement about using Docker Hub instead of the proposed GHCR. 

In summary, there's a missing step and more work needs to be done and there is some confusion between two people as to where to upload the image after it's built.

**Comment by brauliodiasribeiro** [comment]
> > Someone help me install Fabric AI in a docker on Unraid? I don't know how to build this docker in Unraid because it asks for the image path.
> 
> It would be best to learn to build the image yourself. Unless I am mistaken, based on reading this PR, you still need to build the image for yourself because the choice to automate the build and upload to GHCR was pushed back on the core dev, and then there was some confusing statement about using Docker Hub instead of the proposed GHCR.
> 
> In summary, there's a missing step and more work needs to be done and there is some confusion between two people as to where to upload the image after it's built.

I don't have the knowledge to do this. I'll study more and follow the forum. Thanks for your help

**Pull request closed without merging**